process/process.go

```diff
@@ -662,20 +662,39 @@ func (p *Process) tryCommitUponSufficientPrecommits(round Round) {
 // received, or a Precommit is received. Because this method checks whichever
 // Round is relevant (i.e. the Round of the Propose/Prevote/Precommit), and an
 // increase in the current Round can only cause this condition to be closed, it
-// does not need to be tried whenever the current Round changes.
+// does not need to be tried whenever the current Round changes. The f+1
+// messages (one propose and f prevotes/precommits) must be from f+1 unique
+// signatories.
 func (p *Process) trySkipToFutureRound(round Round) {
 	if round <= p.CurrentRound {
 		return
 	}
 
-	msgsInRound := 0
-	if _, ok := p.ProposeLogs[round]; ok {
-		msgsInRound = 1
+	// declare a map to note unique signatories to have sent messages for
+	// round > currentRound
+	uniqueSignatories := make(map[id.Signatory]struct{})
+
+	// handle the propose message
+	propose, ok := p.ProposeLogs[round]
+	if ok {
+		uniqueSignatories[propose.From] = struct{}{}
+	}
+
+	// loop over the prevotes to fetch unique prevoters
+	for signatory := range p.PrevoteLogs[round] {
+		if _, ok := uniqueSignatories[signatory]; !ok {
+			uniqueSignatories[signatory] = struct{}{}
+		}
+	}
+
+	// loop over the precommits to fetch unique precommiters
+	for signatory := range p.PrecommitLogs[round] {
+		if _, ok := uniqueSignatories[signatory]; !ok {
+			uniqueSignatories[signatory] = struct{}{}
+		}
 	}
```

Comment:
This kind of iteration will be very expensive to perform every time we receive a message. Instead, consider adding `TraceLogs map[Round]map[Signatory]bool` to `type State`. Then, whenever a `Propose`, `Prevote`, or `Precommit` is received and accepted (i.e. insert into `ProposeLogs`, `PrevoteLogs`, or `PrecommitLogs`) we should add the signatory to `TraceLogs`. Then, here, we can do `msgsInRound := len(p.TraceLogs[Round])`. We will need to also clear the `TraceLogs` when we progress the height.

Reply:
Yup, I agree this approach looks much better. Just a small difference, I have used `TraceLogs` only for `prevote` and `precommit`, while sticking to `ProposeLogs` for `propose` messages. The reason is only to not keep the `insertPropose` method as it is, since the logic followed there is to overwrite `ProposeLogs` with the most recent valid `propose`. And that would complicate matters further to update `TraceLogs`.

Reply:
Should this have been marked as resolved?